channels/offline-channel-js/src/OfflineChannel.ts

```diff
@@ -0,0 +1,614 @@
+                _setupInMemoTimer();
+                _setSendNextTimer();
+            };
+
+            
+            _self.onunloadFlush = () => {
+                if (!_paused) {
+                    //TODO: or should try send first
+                    let shouldContinue = true;
+                    while (_inMemoBatch.count() && shouldContinue) {
+                        shouldContinue = _flushInMemoItems(true);
+                    }
+                    // TODO: unloadprovider might send events out of order
+                }
+            };
+
+            _self.flush = (sync: boolean, callBack?: (flushComplete?: boolean) => void, sendReason?: SendRequestReason): boolean | void | IPromise<boolean> => {
+                // No op
+            };
+
+            _self.getOfflineListener = () => {
+                return _offlineListener;
+
+            };
+
+            _self._doTeardown = (unloadCtx?: IProcessTelemetryUnloadContext, unloadState?: ITelemetryUnloadState) => {
+                _self.onunloadFlush();
+                _offlineListener && runTargetUnload(_offlineListener, false);
+                let handler = _urlCfg.batchHandler;
+                handler && handler.teardown();
+                _clearScheduledTimer();
+                _initDefaults();
+            };
+
+
+            _self["_getDbgPlgTargets"] = () => {
+                return [_urlCfg, _inMemoBatch, _senderInst];
+            };
+            
+
+            function _initDefaults() {
+                _hasInitialized = false;
+                _paused = false;
+                _sender = new Sender();
+                _urlCfg = null;
+                _offlineListener = null;
+                _diagLogger = null;
+                _endpoint = null;
+                _inMemoBatch = null;
+                _convertUndefined = undefValue;
+                _maxBatchSize = null;
+                _sendNextBatchTimer = null;
+                _consecutiveErrors = null;
+                _retryAt = null;
+                _maxBatchInterval = null;
+                _senderInst = null;
+                _offineSupport = null;
+                _primaryChannelId = null;
+                _overrideIkey = null;
+            }
+
+
+            function _shouldCacheEvent(urlConfig: IUrlLocalStorageConfig, item: IPostTransmissionTelemetryItem) {
+                if ((item.persistence) < urlConfig.minPersistenceCacheLevel) {
+                    return false;
+                }
+
+                return true;
+            }
+
+
+            function _setupInMemoTimer() {
+                if (!_inMemoFlushTimer) {
+                    _inMemoFlushTimer = scheduleTimeout(() => {
+                        _flushInMemoItems();
+                        _inMemoFlushTimer.refresh();
+                        
+                    }, _inMemoTimerOut);
+                    _inMemoFlushTimer.unref();
+                }
+            }
+
+            function _flushInMemoItems(unload?: boolean) {
```

Comment:
Future: as part of the TODO you have on line 173 etc in the `processTelemetry` this should probably return the promise so that it can react (after the batch is stored) that the inMemory batch was added (or not)

Reply:
created task 26681220 for this